### PR TITLE
Feature: Allow pre-reads on fee vault

### PR DESCRIPTION
**Pull request by talekhinezh**
This is a feature required for better substate management. Track layer load state should not get exposed to application layer. 

This break of abstraction prevented allowing the application layer from obtaining locks on it's own since the AuthModule acquires these locks first.

### CODE REVIEW

File: radix-engine/src/engine/track.rs [hunk 1]
```diff
@@ -233,10 +247,13 @@ impl<'s, R: FeeReserve> Track<'s, R> {
...
             LockState::Read(n) => loaded_substate.lock_state = LockState::Read(n - 1),
-            LockState::Write => loaded_substate.lock_state = LockState::no_lock(),
+            LockState::Write => {
+                loaded_substate.lock_state = LockState::no_lock();
+                loaded_substate.metastate = SubstateMetaState::Updated;
+            }
         }
 
-        if write_through {
+        if unmodified_base {
```
Comment (iamyulong) [inline radix-engine/src/engine/track.rs:256]: I think `unmodified_base` only capture one aspect - the substate must pre-exist and unchanged. It does not indicate that any change must be immediately committed.

How about two flags `unmodified | write_through`?

**Review by iamyulong — APPROVED** [review]
Nice improvement!

File: radix-engine/src/engine/track.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine/src/engine/track.rs:256]: hrm good point

Comment (talekhinezh) [reply radix-engine/src/engine/track.rs:256]: what if we called it `force_write` instead

Comment (iamyulong) [reply radix-engine/src/engine/track.rs]: Sounds good.

**Pull request merged**